libs/pika/async_cuda/tests/unit/cuda_future.cu

```diff
@@ -19,22 +19,50 @@
 #include <utility>
 #include <vector>
 
-// -------------------------------------------------------------------------
-// This example uses the normal C++ compiler to compile all the pika stuff
-// but the cuda functions go in their own .cu file and are compiled with
-// nvcc, we don't mix them.
-// Declare functions we are importing - note that template instantiations
-// must be present in the .cu file and compiled so that we can link to them
+namespace cu = pika::cuda::experimental;
+namespace ex = pika::execution::experimental;
+namespace tt = pika::this_thread::experimental;
+
+constexpr auto cuda_launch_kernel = [](auto&&... ts) {
+    cu::check_cuda_error(cudaLaunchKernel(std::forward<decltype(ts)>(ts)...));
+};
+
+constexpr auto cuda_memcpy_async = [](auto&&... ts) {
+    cu::check_cuda_error(cudaMemcpyAsync(std::forward<decltype(ts)>(ts)...));
+};
+
+__global__ void saxpy(int n, float a, float* x, float* y)
+{
+    int i = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i < n)
+        y[i] = a * x[i] + y[i];
+}
+
+template <typename Sender>
+auto launch_saxpy_kernel(pika::cuda::experimental::cuda_scheduler& cuda_sched,
+    Sender&& predecessor, unsigned int& blocks, unsigned int& threads,
+    void** args)
+{
+    return ex::when_all(std::forward<Sender>(predecessor),
+               ex::just(reinterpret_cast<const void*>(&saxpy), dim3(blocks),
+                   dim3(threads), args, std::size_t(0))) |
+        ex::transfer(cuda_sched) | cu::then_with_stream(cuda_launch_kernel);
+}
+
+template <typename T>
+__global__ void trivial_kernel(T val)
+{
+    // TODO: Fingers crossed that printf now works with HIP?
+    printf("hello from gpu with value %f\n", static_cast<double>(val));
```

Comment:
@aurianer do you remember the details of this? `printf` compiles but simply does not print anything with certain HIP configurations, correct?

Reply:
I think I had an error with this, but I can't recall if it was compile or runtime error :/

Reply:
Actually looks like it was a hang:
https://gist.github.com/aurianer/e74acaf9856710c10011c3fb629cb98a

Reply:
At least the HIP build on GitHub actions works, but that doesn't run it of course... I'll see if I can get a HIP environment working for myself again. I'll test it there.

Reply:
It does work with ROCm 5.0. Waiting for builds with 4.5 and 4.3...

Reply:
And it seems to work with 4.3 as well! I'm skipping 4.5 because of the time it takes to build it...